src/webgpu/shader/execution/robust_access_vertex.spec.ts

```diff
@@ -207,181 +267,200 @@ class DrawCall {
     {
       let currAttribute = 0;
-      for (let i = 0; i < bufferContents.length; i++) {
+      for (let i = 0; i < bufferCount; i++) {
         buffers.push({
-          arrayStride: attributesPerBuffer * typeInfo.size,
+          arrayStride: attributesPerBuffer * typeInfo.sizeInBytes,
           stepMode: i === 0 ? 'instance' : 'vertex',
           attributes: Array(attributesPerBuffer)
             .fill(0)
             .map((_, i) => ({
               shaderLocation: currAttribute++,
-              offset: i * typeInfo.size,
-              format: p.type as GPUVertexFormat,
+              offset: i * typeInfo.sizeInBytes,
+              format: type as GPUVertexFormat,
             })),
         });
       }
     }
+    return buffers;
+  }
 
-    // Offset the range checks for gl_VertexIndex in the shader if we use BaseVertex
-    let vertexIndexOffset = 0;
-    if (p.drawCallTestParameter === 'baseVertex') {
-      vertexIndexOffset += p.errorScale;
+  generateVertexShaderCode({
+    bufferCount,
+    attributesPerBuffer,
+    validValues,
+    typeInfo,
+    vertexIndexOffset,
+    numVertices,
+    isIndexed,
+  }: {
+    bufferCount: number;
+    attributesPerBuffer: number;
+    validValues: number[];
+    typeInfo: VertexInfo;
+    vertexIndexOffset: number;
+    numVertices: number;
+    isIndexed: boolean;
+  }): string {
+    // Create layout and attributes listing
+    let layoutStr = 'struct Attributes {';
+    const attributeNames = [];
+    {
+      let currAttribute = 0;
+      for (let i = 0; i < bufferCount; i++) {
+        for (let j = 0; j < attributesPerBuffer; j++) {
+          layoutStr += `[[location(${currAttribute})]] a_${currAttribute} : ${typeInfo.wgslType};\n`;
+          attributeNames.push(`a_${currAttribute}`);
+          currAttribute++;
+        }
+      }
     }
+    layoutStr += '};';
 
-    const pipeline = t.device.createRenderPipeline({
+    const vertexShaderCode: string = `
+      ${layoutStr}
+
+      fn valid(f : f32) -> bool {
+        return ${validValues.map(v => `f == ${v}.0`).join(' || ')};
+      }
+
+      fn validationFunc(v : ${typeInfo.wgslType}) -> bool {
+        ${typeInfo.validationFunc}
+      }
+
+      [[stage(vertex)]] fn main(
+        [[builtin(vertex_index)]] VertexIndex : u32,
+        attributes : Attributes
+        ) -> [[builtin(position)]] vec4<f32> {
+        var attributesInBounds : bool = ${attributeNames
+          .map(a => `validationFunc(attributes.${a})`)
+          .join(' && ')};
+        var indexInBounds : bool = VertexIndex == 0u ||
+            (VertexIndex >= ${vertexIndexOffset}u &&
+            VertexIndex < ${vertexIndexOffset + numVertices}u);
+
+        var Position : vec4<f32>;
+        if (attributesInBounds && (${!isIndexed} || indexInBounds)) {
+          // Success case, move the vertex to the right of the viewport to show that at least one case succeed
+          Position = vec4<f32>(1.0, 0.0, 0.0, 1.0);
```

Comment:
nit: Can we use -0.5 and 0.5 so that we are on the pixel centers?

Reply:
Done. With this change the tests passed on Mac OS Intel GPU. (If using `-1.0` and `1.0` all tests failed)